hilti/toolchain/src/compiler/cfg.cc

```diff
@@ -0,0 +1,893 @@
+    // TODO(bbannier): Make this part of e.g., `UnqualifiedType` instead of a hardcoded list of types here?
+    return type.isA<type::stream::View>() ||     //
+           type.isA<type::bytes::Iterator>() ||  //
+           type.isA<type::list::Iterator>() ||   //
+           type.isA<type::map::Iterator>() ||    //
+           type.isA<type::set::Iterator>() ||    //
+           type.isA<type::stream::Iterator>() || //
+           type.isA<type::vector::Iterator>();
+}
+
+static bool contains(const Node& outer, const Node& inner) {
+    const auto* n = &inner;
+
+    do {
+        if ( n == &outer )
+            return true;
+        n = n->parent();
+    } while ( n );
+
+    return false;
+}
+
+CFG::CFG(const Node* root)
+    : _begin(getOrAddNode(create_meta_node<Start>())), _end(getOrAddNode(create_meta_node<End>())) {
+    assert(root && root->isA<statement::Block>() && "only building from blocks currently supported");
+
+    _begin = addGlobals(_begin, *root);
+    auto* last = addBlock(_begin, root->children(), root);
+    if ( last != _end )
+        addEdge(last, _end);
+}
+
+Node* CFG::addGlobals(Node* parent, const Node& root) {
+    auto* p = root.parent();
+    if ( ! p )
+        return parent;
+
+    auto* mod = p->tryAs<declaration::Module>();
+    if ( ! mod )
+        return parent;
+
+    // A global variables with an init statement since they are effectively statements.
+    for ( auto* decl : mod->declarations() ) {
+        auto* global = decl->tryAs<declaration::GlobalVariable>();
+        if ( ! global )
+            continue;
+
+        if ( ! global->init() )
+            continue;
+
+        auto* stmt = getOrAddNode(global);
+        addEdge(parent, stmt);
+        parent = stmt;
+    }
+
+    return parent;
+}
+
+Node* CFG::addBlock(Node* parent, const Nodes& stmts, const Node* scope) {
+    // If `children` directly has any statements which change control flow like
+    // `throw` or `return` any statements after that are unreachable. To model
+    // such ASTs we add a flow with all statements up to the "last" semantic
+    // statement (either the last child or the control flow statement) to the
+    // CFG under `parent`. Statements after that are added as children without
+    // parents, and mixed with the previous flow.
+
+    // After this block `last` is the last reachable statement, either end of
+    // children or a control flow statement.
+    auto last = std::find_if(stmts.begin(), stmts.end(), [](auto&& c) {
+        return c && (c->template isA<statement::Return>() || c->template isA<statement::Throw>() ||
+                     c->template isA<statement::Continue>() || c->template isA<statement::Break>());
+    });
+    const bool has_dead_flow = last != stmts.end();
+    if ( has_dead_flow )
+        last = std::next(last);
+
+    // Node this block will eventually flow into.
+    auto* scope_end = getOrAddNode(create_meta_node<ScopeEnd>(scope));
```

Comment:
(not sure exactly where this comment applies)

Scoping for blocks gets into a similar problem to lambda captures, so this code will incorrectly remove a dead store of `x = 5;`:

```
function void f() {
    local uint<64> x = 1;
    { x = 5; local uint<64> x = 10; }
    hilti::print(x);
}
```

Generally the solution would be to add a counter "how many scopes out until I found `x`?" - I guess that'd be a de Bruijn index

Though some other stuff could work too, with symbol tables attached to the scopes or transfers or something (I haven't really thought much about it)

Reply:
This should be fixed now, in the existing framework.

Since we always uses by `Name` which in valid code always resolves to exactly one declaration there should be no issue with us confusing the outer and the inner `x`. I added a test to make sure things stay that way.